lib/rb-inotify/notifier.rb

```diff
@@ -51,9 +94,13 @@ def self.supports_ruby_io?
     # @return [Notifier]
     # @raise [SystemCallError] if inotify failed to initialize for some reason
     def initialize
-      @fd = Native.inotify_init
+      fd = Native.inotify_init
       @watchers = {}
-      return unless @fd < 0
+      unless fd < 0
+        @handle = Handle.new(fd)
+        ObjectSpace.define_finalizer(self, @handle.method(:close))
```

Comment:
@matthewd I'm pretty sure that this line exhibits this bug:

http://www.mikeperham.com/2010/02/24/the-trouble-with-ruby-finalizers/

A simple example seems to confirm this:

``` ruby
# notifier.rb
class Notifier
  class Handle
    def close
      puts "Close Handle: #{object_id}"
    end
  end

  def initialize
    @handle = Handle.new
    puts "Open Handle: #{@handle.object_id}"
    ObjectSpace.define_finalizer(self, @handle.method(:close))
  end
end

n = Notifier.new
n = nil

GC.start
sleep 1
````

``` sh
$ ruby notifier.rb 
Open Handle: 70347030523800
```

Changing to use a class method as suggested in the article gives the following:

``` ruby
# notifier.rb
class Notifier
  class Handle
  end

  def self.close_handle(handle)
    proc { puts "Close Handle: #{handle.object_id}" }
  end

  def initialize
    @handle = Handle.new
    puts "Open Handle: #{@handle.object_id}"
    ObjectSpace.define_finalizer(self, self.class.close_handle(@handle))
  end
end

n = Notifier.new
n = nil

GC.start
sleep 1
```

``` sh
$ ruby notifier.rb 
Open Handle: 70250977331740
Close Handle: 70250977331740
```

The PR does fix the problem we were seeing in Rails, but only because I think it's preventing the garbage collector from cleaning up the object due to the reference to self :smile: